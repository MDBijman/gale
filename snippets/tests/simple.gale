main: ui8 -> ui8
a = let
      d: [ui8; 3] = [1, 2, 3],
      c: (\b -> b) b
    in
      c a.



