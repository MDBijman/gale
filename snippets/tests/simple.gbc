mod simple

.code
fn fib ($0: ui64) -> ui64
    mov $1, $0
    movc $2, 0
    eq $3, $1, $2
    mov $4, $3
    neg $5, $4
    mov $6, $5
    jmpif $6 @lbl_1
    movc $7, 1
    jmp @lbl_0
  lbl_1:
    mov $8, $0
    movc $9, 1
    eq $10, $8, $9
    mov $11, $10
    neg $12, $11
    mov $13, $12
    jmpif $13 @lbl_3
    movc $14, 1
    jmp @lbl_2
  lbl_3:
    movc $15, 1
    neg $16, $15
    mov $17, $16
    jmpif $17 @lbl_4
    mov $18, $0
    mov $19, $18
    movc $20, 1
    sub $21, $19, $20
    mov $22, $21
    call $23, @fib ($22)
    mov $24, $23
    mov $25, $18
    movc $26, 2
    sub $27, $25, $26
    mov $28, $27
    call $29, @fib ($28)
    mov $30, $29
    add $31, $24, $30
    mov $32, $31
  lbl_4:
    mov $14, $32
  lbl_2:
    mov $7, $14
  lbl_0:
    ret $7
endfn

fn main ($0: &[&str]) -> ui64
    movc $1, 1
    neg $2, $1
    mov $3, $2
    jmpif $3 @lbl_5
    mov $4, $0
    mov $5, $4
    movc $6, 0
    idx $7, $5, $6, &str
    mov $8, $7
    call $9, @std:parse_ui64 ($8)
    mov $10, $9
    call $11, @fib ($10)
    mov $12, $11
  lbl_5:
    ret $12
endfn
