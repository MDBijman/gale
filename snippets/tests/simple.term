File([
    FnDecl(
      "main"
    , FnType(
        TypeId("ui8")
      , TypeId("ui8")
      )
    )
  , FnDef(
      "main"
    , "a"
    , Let(
        [
          Decl(
            "d"
          , TypeArray(
              TypeId("ui8")
            , Int(3)
            )
          , Array([
              Int(1)
            , Int(2)
            , Int(3)
            ])
          )
        , Decl(
            "c"
          , TypeId("ui8")
          , Int(2)
          )
        ]
      , Appl(
          Ref("id")
        , Ref("c")
        )
      )
    )
  , FnDecl(
      "id"
    , FnType(
        TypeId("ui8")
      , TypeId("ui8")
      )
    )
  , FnDef(
      "id"
    , "a"
    , Ref("a")
    )
  ])