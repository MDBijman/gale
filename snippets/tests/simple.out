File([Function(
    "main"
  , Meta(
      InSize(1)
    , OutSize(1)
    )
  , Scope(
      Size(9)
    , [
        Store(
          "d"
        , Array([
            UI8(1)
          , UI8(2)
          , UI8(3)
          ])
        )
      , Store(
          "c"
        , UI8(1)
        )
      , Load("c")
      ]
    )
  )])